std/bitmanip.d

```diff
@@ -1710,6 +1710,242 @@ public:
+     * boundary may be overwritten by this operation. It does not attempt to
+     * preserve bits past the end of the array.)
+     */
+    void opOpAssign(string op)(size_t nbits) @nogc pure nothrow
+        if (op == ">>")
+    {
+        size_t wordsToShift = nbits / bitsPerSizeT;
+        size_t bitsToShift = nbits % bitsPerSizeT;
+
+        if (wordsToShift < dim)
+        {
+            // Unfortunately, due to the way core.bitop.bt() works, indices are
+            // the reverse order of what an equivalent >> on a size_t would do,
+            // so we have to use rollLeft() instead of rollRight() here,
+            // contrary to the direction of the >> operator on this BitArray.
+            foreach_reverse (i; 1 .. dim - wordsToShift)
+            {
+                ptr[i + wordsToShift] = rollLeft(ptr[i], ptr[i-1],
+                                                 bitsToShift);
+            }
+            ptr[wordsToShift] = rollLeft(ptr[0], 0, bitsToShift);
+        }
+
+        import std.algorithm : min;
+        foreach (i; 0 .. min(wordsToShift, dim))
+        {
+            ptr[i] = 0;
+        }
+    }
+
+    /**
+     * Operator $(D <<=) support.
+     *
+     * Shifts all the bits in the array to the left by the given number of
+     * bits.  The leftmost bits are dropped, and 0's are appended to the end to
+     * fill up the vacant bits.
+     *
+     * $(RED Warning: unused bits in the final word up to the next word
+     * boundary may be overwritten by this operation. It does not attempt to
+     * preserve bits past the end of the array.)
+     */
+    void opOpAssign(string op)(size_t nbits) @nogc pure nothrow
+        if (op == "<<")
+    {
+        size_t wordsToShift = nbits / bitsPerSizeT;
+        size_t bitsToShift = nbits % bitsPerSizeT;
+
+        // Unfortunately, due to the way core.bitop.bt() works, indices are the
+        // reverse order of what an equivalent << on a size_t would do, so we
+        // have to use rollRight() instead of rollLeft() here, contrary to the
+        // direction of the << operator on this BitArray.
+        if (wordsToShift + 1 < dim)
+        {
+            foreach (i; 0 .. dim - wordsToShift - 1)
+            {
+                ptr[i] = rollRight(ptr[i + wordsToShift + 1],
+                                   ptr[i + wordsToShift], bitsToShift);
+            }
+        }
+
+        // The last word needs some care, as it must shift in 0's from past the
+        // end of the array.
+        if (wordsToShift < dim)
+        {
+            ptr[dim - wordsToShift - 1] = rollRight(0, ptr[dim - 1] & endMask,
+                                                    bitsToShift);
+        }
+
+        import std.algorithm : min;
+        foreach (i; 0 .. min(wordsToShift, dim))
+        {
+            ptr[dim - i - 1] = 0;
+        }
+    }
+
+    unittest
+    {
+        import std.format : format;
+
+        BitArray b;
+        b.init([1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1]);
```

Comment:
Good point, I'll fix that. I'm surprised a randomly-typed binary sequence came out palindromic. :-P